skxray/cdi.py

```diff
@@ -119,9 +142,284 @@ def convolution(array1, array2):
+    sample_obj[index_v] = 0
+    return sample_obj
+
+
+def cal_relative_error(x_old, x_new):
+    """
+    Relative error is calculated as the ratio of the difference between the new and
+    the original arrays to the norm of the original array.
+
+    Parameters
+    ----------
+    x_old : array
+        previous data set
+    x_new : array
+        new data set
+
+    Returns
+    -------
+    float :
+        relative error
+    """
+    return np.linalg.norm(x_new - x_old)/np.linalg.norm(x_old)
+
+
+def cal_diff_error(sample_obj, diff_array):
+    """
+    Calculate the error in q space.
+
+    Parameters
+    ----------
+    sample_obj : array
+        sample data
+    diff_array : array
+        diffraction pattern
+
+    Returns
+    -------
+    float :
+        relative error in q space
+    """
+    new_diff = np.abs(_fft_helper(sample_obj)) / np.sqrt(np.size(sample_obj))
+    return cal_relative_error(diff_array, new_diff)
+
+
+class CDI(object):
+
+    def __init__(self, diffamp, **kwargs):
+        """
+        Parameters
+        ----------
+        deffmap : array
+            diffraction pattern from experiments
+        kwargs : dict
+            parameters related to cdi reconstruction
+        """
+
+        self.diff_array = np.array(diffamp)     # diffraction data
+
+        self.ndim = self.diff_array.ndim    # 2D or 3D case
+        if self.ndim == 2:
+            self.nx, self.ny = np.shape(self.diff_array)    # array dimension
+        if self.ndim == 3:
+            self.nx, self.ny, self.nz = np.shape(self.diff_array)    # array dimension
+
+        self.pi_modulus_flag = kwargs['pi_modulus_flag']  # 'Complex'
+
+        self.beta = kwargs['beta']  # feedback parameter for difference map algorithm, around 1.15
+        self.start_ave = kwargs['start_ave']  # 0.8
+
+        # initial support
+        self.init_obj_flag = kwargs['init_obj_flag']
+        self.init_sup_flag = kwargs['init_sup_flag']
+        self.sup_radius = kwargs['support_radius']
+        self.sup_shape = kwargs['support_shape']  # 'Box' or 'Disk'
+
+        # parameters for shrink wrap
+        self.shrink_wrap_flag = kwargs['shrink_wrap_flag']
+        self.sw_sigma = kwargs['sw_sigma']  # 0.5
+        self.sw_threshold = kwargs['sw_threshold']  # 0.1
+        self.sw_start = kwargs['sw_start']  # 0.2
+        self.sw_end = kwargs['sw_end']  # 0.8
+        self.sw_step = kwargs['sw_step']  # 10
+
+    def init_obj(self):
+        """
+        Initiate phase. Focus on 2D case here, and to be extended.
+        """
+        pha_tmp = np.random.uniform(0, 2*np.pi, (self.nx, self.ny))
+        self.obj = (_fft_helper(self.diff_array * np.exp(1j*pha_tmp), option='ifft')
+                    * np.sqrt(np.size(self.diff_array)))
+
+    def init_sup(self):
+        """ Initiate sample support.
+        """
+        self.sup = np.zeros_like(self.diff_array)
+        if self.ndim == 2:
+            if self.sup_shape == 'Box':
```

Comment:
OK, I am thinking to break this class into small pieces. In the original code from beamline scientists, all the functions inside this file are included in a single class, causing me lots of headache...